File: 剑指Offer/37.序列化二叉树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// 使用先序遍历加特殊符号的方法来处理；
class Codec {
public:
    void rserialize(TreeNode* root, string& str) {
        if (root == NULL) {
            str += "null,";
            return;
        }

        str += to_string(root->val) + ',';
        rserialize(root->left, str);
        rserialize(root->right, str);
    }

    TreeNode* rdeserialize(string data, int& index) {
        if (index == data.length()) return NULL;
        int k = index;
        while (data[k] != ',') k++;
        if (data[index] == 'n') {
            index = k + 1;
            return NULL;
        }
        
        int val = 0;
        int negative = 1;
        if (index < k && data[index] == '-') {
            // 此节点的值为负数
            negative = -1;
            index++;
        }
        for (int i = index; i < k; i++) {
            val = val * 10 + (data[i] - '0');
        }
        val *= negative;
        index = k + 1;
        TreeNode *root = new TreeNode(val);
        root->left = rdeserialize(data, index);
        root->right = rdeserialize(data, index);
        return root;
    }

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string res = "";
        rserialize(root, res);
        return res;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int index = 0;
        return rdeserialize(data, index);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: 剑指Offer/37.序列化二叉树.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def rserialize(root: TreeNode, s: str) -> str:
            if root == None:
                s += "None,"
            else:
                s += str(root.val) + ','
                s = rserialize(root.left, s)
                s = rserialize(root.right, s)
            return s
        
        return rserialize(root, "")

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def rdeserialize(data: str) -> TreeNode:
            if data[0] == "None":
                data.pop(0)
                return None
            
            curNode = TreeNode(int(data[0]))
            data.pop(0)
            curNode.left = rdeserialize(data)
            curNode.right = rdeserialize(data)
            return curNode

        data = data.split(',')
        root = rdeserialize(data)
        return root

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))